lucene/core/src/java/org/apache/lucene/util/bkd/DocIdsWriter.java

```diff
@@ -24,89 +24,73 @@
 import org.apache.lucene.util.DocBaseBitSetIterator;
 import org.apache.lucene.util.FixedBitSet;
 
-class DocIdsWriter {
+final class DocIdsWriter {
 
-  private DocIdsWriter() {}
+  private static final byte CONTINUOUS_IDS = (byte) -2;
+  private static final byte BITSET_IDS = (byte) -1;
+  private static final byte DELTA_FOR_UTIL = (byte) 32 + 16;
+  private static final byte BPV_24 = (byte) 32 + 24;
+  private static final byte BPV_32 = (byte) 32;
+  // These signs are legacy, should no longer be used in the writing side.
+  private static final byte LEGACY_DELTA_VINT = (byte) 0;
+  private static final byte LEGACY_BPV_24 = (byte) 24;
 
-  static void writeDocIds(int[] docIds, int start, int count, DataOutput out) throws IOException {
+  private final BKDForUtil forUtil;
+  private final int[] scratch;
+
+  DocIdsWriter(int maxPointsInLeaf) {
+    scratch = new int[maxPointsInLeaf];
+    forUtil = new BKDForUtil(maxPointsInLeaf);
+  }
+
+  void writeDocIds(int[] docIds, int start, int count, DataOutput out) throws IOException {
     // docs can be sorted either when all docs in a block have the same value
     // or when a segment is sorted
-    boolean sorted = true;
     boolean strictlySorted = true;
+    int min = docIds[0];
+    int max = docIds[0];
     for (int i = 1; i < count; ++i) {
       int last = docIds[start + i - 1];
       int current = docIds[start + i];
-      if (last > current) {
-        sorted = strictlySorted = false;
-        break;
-      } else if (last == current) {
+      if (last >= current) {
         strictlySorted = false;
       }
+      min = Math.min(min, current);
+      max = Math.max(max, current);
     }
 
-    int min2max = docIds[start + count - 1] - docIds[start] + 1;
+    int min2max = max - min + 1;
     if (strictlySorted) {
       if (min2max == count) {
         // continuous ids, typically happens when segment is sorted
-        out.writeByte((byte) -2);
+        out.writeByte(CONTINUOUS_IDS);
         out.writeVInt(docIds[start]);
         return;
       } else if (min2max <= (count << 4)) {
         assert min2max > count : "min2max: " + min2max + ", count: " + count;
         // Only trigger bitset optimization when max - min + 1 <= 16 * count in order to avoid
         // expanding too much storage.
         // A field with lower cardinality will have higher probability to trigger this optimization.
-        out.writeByte((byte) -1);
+        out.writeByte(BITSET_IDS);
         writeIdsAsBitSet(docIds, start, count, out);
         return;
       }
     }
-    if (sorted) {
-      out.writeByte((byte) 0);
-      int previous = 0;
-      for (int i = 0; i < count; ++i) {
-        int doc = docIds[start + i];
-        out.writeVInt(doc - previous);
-        previous = doc;
+
+    if (Integer.toUnsignedLong(min2max) <= 0xFFFFL) {
```

Comment:
since doc IDs are integers in [0, MAX_VALUE) we don't need to convert to an unsigned long (MAX_VALUE is not a legal doc ID)
```suggestion
    if (min2max <= 0xFFFF) {
```